### PR TITLE
[RISCV]Add support for resolving encoding conflicts among vendor specific CSRs

### PR DESCRIPTION
**Pull request by quic-garvgupt**
This patch adds the framework for resolving encoding conflicts among CSRs.

Specifically, this patch adds a support for emitting a second lookup function for the primary key which takes an additional arguement `List` of type `std::vector` and inside the function definition, will populate the `List` with all sysreg that matches primary key.

While printing the CSR name during objdump, iterate over the `List` and print the name of only that CSR which satisfies the feature requirement of subtarget.

Below are the signatures of the functions for primary key:
```
`const SysReg *lookupSysRegByEncoding(uint16_t Encoding);` 
`void lookupSysRegByEncoding(uint16_t Encoding, std::vector<const SysReg*> &List);`
```

Below is the definition for the second primary function:

```
void lookupSysRegByEncoding(uint16_t Encoding, std::vector<const SysReg*> &List) {
  struct KeyType {
    uint16_t Encoding;
  };
  KeyType Key = {Encoding};
  auto Table = ArrayRef(SysRegsList);
  auto Idx = std::lower_bound(Table.begin(), Table.end(), Key,
    [](const SysReg &LHS, const KeyType &RHS) {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    });

  if (Idx == Table.end() ||
      Key.Encoding != Idx->Encoding)
    return;

  auto UpperBound = std::upper_bound(Table.begin(), Table.end(), Key,
    [](const KeyType &LHS, const SysReg &RHS) {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    });

  while(Idx != UpperBound){
    List.push_back(&SysRegsList[Idx - Table.begin()]);
    Idx++;
  }
}
```

**Usage**: CSRs with same encodings need to be under separate features. Below is example illustrating the same

```
let FeaturesRequired = [{ {Feature1} }] in {
def : SysReg<"csr1", 0x123>;
}

let FeaturesRequired = [{ {Feature2} }] in {
def : SysReg<"csr2", 0x123>;
}
```

### CODE REVIEW

**Comment by quic-garvgupt** [comment]
@topperc @apazos @asb, can you please add yourself as reviewers for this PR? Thanks

File: llvm/utils/TableGen/SearchableTableEmitter.cpp [hunk 1]
```diff
@@ -540,10 +574,20 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
...
   // search can be performed by "Thing".
-  if (Table.PrimaryKey)
-    emitLookupFunction(Table, *Table.PrimaryKey, true, OS);
+  if (Table.PrimaryKey) {
+    auto &Index = *Table.PrimaryKey;
+    // Two lookupfunction functions need to be generated to allow more than one
+    // lookup signature for the primary key lookup : first will return a SysReg
+    // that matches the primary key, second will populate a vector passed as
+    // argument with all the SysRegs that match the primary key.
+    emitLookupFunction(Table, Index, /*IsPrimary=*/true,
```
Comment (wangpc-pp) [inline llvm/utils/TableGen/SearchableTableEmitter.cpp:583]: I think these SearchableTable changes should be in separate PR.

**Comment by quic-garvgupt** [comment]
Abandoning this PR because directly created it for the main branch. Will create a new PR.

Apologies for the inconvenience
